OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyMapPreviewLogic.cs

```diff
@@ -22,100 +24,39 @@ public class LobbyMapPreviewLogic : ChromeLogic
-						m => WidgetUtils.TruncateText("Created by {0}".F(lobby.Map.Author), authorLabel.Bounds.Width, font));
-					authorLabel.GetText = () => author.Update(lobby.Map);
-				}
+				available.IsVisible = () => getMap().Status == MapStatus.Available && (!getMap().RulesLoaded || !getMap().InvalidCustomRules);
+				SetupWidgets(available, getMap, onMouseDown, getSpawnOccupants);
 			}
 
 			var invalid = widget.GetOrNull("MAP_INVALID");
 			if (invalid != null)
 			{
-				invalid.IsVisible = () => lobby.Map.Status == MapStatus.Available && lobby.Map.InvalidCustomRules;
-
-				var preview = invalid.Get<MapPreviewWidget>("MAP_PREVIEW");
-				preview.Preview = () => lobby.Map;
-				preview.OnMouseDown = mi => LobbyUtils.SelectSpawnPoint(orderManager, preview, lobby.Map, mi);
-				preview.SpawnOccupants = () => LobbyUtils.GetSpawnOccupants(orderManager.LobbyInfo, lobby.Map);
-
-				var titleLabel = invalid.GetOrNull<LabelWidget>("MAP_TITLE");
-				if (titleLabel != null)
-					titleLabel.GetText = () => lobby.Map.Title;
-
-				var typeLabel = invalid.GetOrNull<LabelWidget>("MAP_TYPE");
-				if (typeLabel != null)
-				{
-					var type = new CachedTransform<MapPreview, string>(m => lobby.Map.Categories.FirstOrDefault() ?? "");
-					typeLabel.GetText = () => type.Update(lobby.Map);
-				}
+				invalid.IsVisible = () => getMap().Status == MapStatus.Available && getMap().InvalidCustomRules;
+				SetupWidgets(invalid, getMap, onMouseDown, getSpawnOccupants);
 			}
 
 			var download = widget.GetOrNull("MAP_DOWNLOADABLE");
 			if (download != null)
 			{
-				download.IsVisible = () => lobby.Map.Status == MapStatus.DownloadAvailable;
-
-				var preview = download.Get<MapPreviewWidget>("MAP_PREVIEW");
-				preview.Preview = () => lobby.Map;
-				preview.OnMouseDown = mi => LobbyUtils.SelectSpawnPoint(orderManager, preview, lobby.Map, mi);
-				preview.SpawnOccupants = () => LobbyUtils.GetSpawnOccupants(orderManager.LobbyInfo, lobby.Map);
-
-				var titleLabel = download.GetOrNull<LabelWidget>("MAP_TITLE");
-				if (titleLabel != null)
-					titleLabel.GetText = () => lobby.Map.Title;
-
-				var typeLabel = download.GetOrNull<LabelWidget>("MAP_TYPE");
-				if (typeLabel != null)
-				{
-					var type = new CachedTransform<MapPreview, string>(m => lobby.Map.Categories.FirstOrDefault() ?? "");
-					typeLabel.GetText = () => type.Update(lobby.Map);
-				}
-
-				var authorLabel = download.GetOrNull<LabelWidget>("MAP_AUTHOR");
-				if (authorLabel != null)
-					authorLabel.GetText = () => "Created by {0}".F(lobby.Map.Author);
+				download.IsVisible = () => getMap().Status == MapStatus.DownloadAvailable;
+				SetupWidgets(download, getMap, onMouseDown, getSpawnOccupants);
 
 				var install = download.GetOrNull<ButtonWidget>("MAP_INSTALL");
 				if (install != null)
 				{
-					install.OnClick = () => lobby.Map.Install(mapRepository, () =>
+					install.OnClick = () => getMap().Install(mapRepository, () =>
```

Comment:
This is a good example of the `getMap()` danger: it generally takes between half and a few seconds to install a map, and its quite possible for the host to change map between `getMap().Install` and `getMap().PreloadRules` running.

The fix again is to evaluate it once to be certain that it can't change under you:
```csharp
install.OnClick = () => 
{
	var map = getMap();
	map.Install(mapRepository, () =>
	{
		map.PreloadRules();
		if (orderManager != null)
			Game.RunAfterTick(() => orderManager.IssueOrder(Order.Command("state {0}".F(Session.ClientState.NotReady))));
	}
};
```